server/channels/web/context.go

```diff
@@ -139,6 +143,119 @@ func (c *Context) SessionRequired() {
+		!*license.Features.CustomTermsOfService {
+		return nil
+	}
+
+	// Exempt bot accounts - bots are automated systems and do not need to accept ToS
+	session := c.AppContext.Session()
+	if session == nil {
+		return model.NewAppError("TermsOfServiceRequired", "api.context.session_expired.app_error", nil, "Session is nil", http.StatusUnauthorized)
+	}
+	if session.IsBotUser() {
+		return nil
+	}
+
+	// Get latest ToS document - fail closed on database errors, fail open if no ToS exists
+	latestToS, err := c.App.Srv().Store().TermsOfService().GetLatest(true)
+	if err != nil {
+		// Check if this is specifically "no ToS exists" vs database error
+		var nfErr *store.ErrNotFound
+		if errors.As(err, &nfErr) {
+			c.Logger.Error("No Terms of Service document found - ToS enforcement disabled system-wide", mlog.String("context", "TermsOfServiceRequired"))
+			return nil // No ToS exists, allow access
+		}
+		// Database error - fail closed for security
+		c.Logger.Error("Database error retrieving terms of service", mlog.Err(err))
+		return model.NewAppError("TermsOfServiceRequired", "api.context.terms_of_service_required.app_error", nil, "", http.StatusForbidden)
+	}
+
+	// Check session cache for user's ToS acceptance
+	if acceptedToSId, exists := session.Props[model.SessionPropTermsOfServiceId]; exists {
+		if acceptedToSId == latestToS.Id {
+			return nil // User already accepted latest ToS
+		}
+	}
+
+	// Cache miss - check database for user's ToS acceptance
+	userToS, appErr := c.App.GetUserTermsOfService(session.UserId)
+	if appErr != nil {
+		if appErr.StatusCode == http.StatusNotFound {
+			// User hasn't accepted ToS yet
+			return model.NewAppError("TermsOfServiceRequired", "api.context.terms_of_service_required.app_error", nil, "", http.StatusForbidden)
+		}
+		c.Logger.Error("Error retrieving user terms of service", mlog.String("user_id", session.UserId), mlog.Err(appErr))
+		return model.NewAppError("TermsOfServiceRequired", "api.context.terms_of_service_required.app_error", nil, "", http.StatusForbidden)
+	}
+
+	// Check if user accepted latest ToS
+	if userToS.TermsOfServiceId != latestToS.Id {
+		return model.NewAppError("TermsOfServiceRequired", "api.context.terms_of_service_required.app_error", nil, "", http.StatusForbidden)
+	}
+
+	// Cache the acceptance in session for future requests
+	session.AddProp(model.SessionPropTermsOfServiceId, userToS.TermsOfServiceId)
+
+	return nil
+}
+
+// isExactPathMatch checks if path exactly matches any of the provided exempt paths
+// Uses explicit path cleaning to prevent bypass attempts via path traversal
+func isExactPathMatch(requestPath string, exemptPaths []string) bool {
+	cleanPath := path.Clean(requestPath)
+	return slices.Contains(exemptPaths, cleanPath)
+}
+
+// matchesUserToSPattern checks if path matches the pattern /api/v4/users/{user_id}/terms_of_service
+// using the same validation pattern as Gorilla Mux: {user_id:[A-Za-z0-9]+}
```

Comment:
Having to copy the Gorilla Mux validation pattern suggests to me this logic belongs in `api4/` not `web/`.

Can we extend the `api.APISessionRequired` with an option to flag those APIs that can always be used when TOS is enforced? See `RequireMfa` as a comparable.